### ISSUE TITLE
PublicFolder.GetFilePath() may be vulnerable to directory traversal attack

### ISSUE THREAD

**Issue by ejhg** [sukona/Grapevine#127]
PublicFolder.GetFilePath() should probably include a check to prevent directory traversal attacks.

This is the original code:

```
        /// <summary>
        /// Returns the path to the file specified by the pathinfo
        /// </summary>
        private string GetFilePath(string pathinfo)
        {
            if (string.IsNullOrWhiteSpace(pathinfo)) return null;

            var path = Path.Combine(FolderPath,
                (string.IsNullOrWhiteSpace(Prefix) ? pathinfo : pathinfo.Replace(Prefix, "")).TrimStart('/', '\\')
                    .Replace("/", Path.DirectorySeparatorChar.ToString()));

            if (File.Exists(path)) return path;
            if (!Directory.Exists(path)) return null;

            path = Path.Combine(path, DefaultFileName);
            return File.Exists(path) ? path : null;
        }
```

This is a simple possible patch:

```
if (File.Exists (path)) {
    // Protect against directory traversal attacks
    fileDirectory = new FileInfo (path).Directory.FullName;
    webRootDirectory = new Directory(webRootPath).FullName;
    if (fileDirectory.StartsWith (webRootDirectory, StringComparison.InvariantCulture)) {
        return path;
    }
}
```

Comments:
- The web root path would need to be available in this context
- Perhaps there's a better way to compare directories than the one I propose.

**Comment by ejhg** [comment]
I can create a pull request if needed.

**Comment by ejhg** [comment]
Note: I have not been able to execute an attack, however.  It is possible that HttpListener is sanitizing the URL, but I can't find any docs on that.

I have verified that PublicFolder.GetFilePath() itself is vulnerable to the attack, but I have not been able to make a malicious filename reach this method.

**Comment by scottoffen** [comment]
It may or may be a vulnerability, but I'd prefer to err on the side of caution. If you have a fix for it, go ahead and submit a PR. Make sure it still passes all unit tests, and add a unit test for your fix as well.

Thanks!

**Comment by scottoffen** [comment]
Vulnerability or not, this will be resolved with the proposed changes in issue [#79](https://github.com/sukona/Grapevine/issues/79)

**Comment by scottoffen** [comment]
This is (or will) no longer an issue with the folder content caching feature introduced.
